src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs

```diff
@@ -76,6 +79,13 @@ public EndpointHtmlRenderer(IServiceProvider serviceProvider, ILoggerFactory log
         // (which will obviously not work, but should not fail)
         var componentApplicationLifetime = httpContext.RequestServices.GetRequiredService<ComponentStatePersistenceManager>();
         await componentApplicationLifetime.RestoreStateAsync(new PrerenderComponentApplicationStore());
+
+        if (componentType != null)
+        {
+            // Saving RouteData to avoid routing twice in Router component
+            var routingStateProvider = httpContext.RequestServices.GetService<RoutingStateProvider>();
+            ((EndpointRoutingStateProvider)routingStateProvider!).SetRouteData(new RouteData(componentType, httpContext.GetRouteData().Values));
```

Comment:
This cast seems unsafe since `RoutingStateProvider` is public and you're using `TryAddScoped` which implies it could be overwritten by the user. I think even with `AddScoped` it could still be replaced, but don't quote me on that.

Reply:
Thanks @halter73! I registered `EndpointRoutingStateProvider` directly now.

> But if we're going that route (no pun intended), why not use an IRoutingStateProvider interface over a base class?

We certainly can do that. The downside to using `IRoutingStateProvider` over the base class `RoutingStateProvider` would be that in all the other places we would have same implementation of IRoutingStateProvider which is `RouteData` getter returns null. What do we usually do in these types of situations?

Reply:
We have quite a few internal service implementations that are just no-ops because they are not relevant in a particular hosting environment. There wouldn't be anything bad about having a `WebAssemblyRoutingStateProvider` that just always returns `null` with a comment saying it's not relevant in this environment.

Alternatively, you could change `Router` so that it no longer has an `[Inject]` for the routing state provider, but instead gets it using `serviceProvider.GetService<T>()`, so that it's an optional dependency. Environments that don't supply an implementation of this service will just give `null`. Then it's only necessary to register an implementation in the endpoints case.

Reply:
Thanks @SteveSandersonMS ! I'd better do that then.

Reply:
I tried the second option and it didn't work. The reason is `serviceProvider.GetService<T>()` doesn't return null if the service wasn't registered. It throws InvalidOperationException.

If I go with the first option then I'll need new API review. I think I'm going to wait for @javiercn to come back before doing that change since he has already approved the API with class `RoutingStateProvider`.

Reply:
> The reason is serviceProvider.GetService<T>() doesn't return null if the service wasn't registered. It throws InvalidOperationException.

Are you sure? `GetService` is explicitly documented as returning `null` in this case: https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceproviderserviceextensions.getservice?view=dotnet-plat-ext-8.0#returns

There is a separate API, `GetRequiredService`, for cases where it should throw when no implementation was registered.

Reply:
Since I don't need `RoutingStateProvider` that returns null any more, I'm going to use interface `IRoutingStateProvider`.